pipeline/pipeline.go

```diff
@@ -0,0 +1,182 @@
+package pipeline
+
+import (
+	"bytes"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/fullstorydev/hauser/config"
+	"github.com/nishanths/fullstory"
+)
+
+// A Pipeline downloads data exports from fullstory.com and saves them to local disk
+type Pipeline struct {
+	transformRecord func(map[string]interface{}) ([]string, error)
+	metaTime        time.Time
+	metaCh          chan fullstory.ExportMeta
+	expCh           chan ExportData
+	saveCh          chan SavedExport
+	recsCh          chan RecordGroup
+	errCh           chan error
+	conf            *config.Config
+}
+
+// NewPipeline returns a new Pipeline base on the configuration provided
+func NewPipeline(conf *config.Config, transformRecord func(map[string]interface{}) ([]string, error)) *Pipeline {
+	return &Pipeline{
+		conf:            conf,
+		transformRecord: transformRecord,
+		metaCh:          make(chan fullstory.ExportMeta),
+		expCh:           make(chan ExportData),
+		recsCh:          make(chan RecordGroup),
+		saveCh:          make(chan SavedExport),
+		errCh:           make(chan error),
+	}
+}
+
+// Start begins pipeline downloading and processing bundles after the provided time. This function returns a channel
+// that will contain the filenames to which the data was saved and a channel that contains any errors that occur in the
+// pipeline. These must be retrieved from the channel to continue processing.
+func (p *Pipeline) Start(startTime time.Time) (chan SavedExport, chan error) {
+	p.metaTime = startTime
+	go p.metaFetcher()
```

Comment:
I think our goal is that if an error occurs at any stage of the pipeline, all stages shut down. However, if an error occurs in a later stage of the pipeline, this goroutine will never be shut down. The other goroutines are reading from a channel, so closing their upstream channel causes them to finish when they get a stop signal from a channel. This one isn't listening on a channel, it's just fetching exports, so it will only quit when there are no more exports available.

In this case you could have a second channel for communicating an early stop, or you could also use a context with a cancel function, which has some support for this bult-in: https://blog.golang.org/context.

Reply:
I had that at one point, but had removed it when I was cleaning things up and realized that the channels should close when they are using `range`. Clearly I forgot the first one 🤦‍♂